File: src/ros2_bridge/mappers/visualization_msgs/interactive_marker_update.rs
//! Typed mapper for `visualization_msgs/msg/InteractiveMarkerUpdate`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn interactive_marker_update_to_bus(msg: ros_env::visualization_msgs::msg::InteractiveMarkerUpdate) -> crate::visualization_msgs::msg::v1::InteractiveMarkerUpdate {
    crate::visualization_msgs::msg::v1::InteractiveMarkerUpdate {
        server_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.server_id),
        seq_num: msg.seq_num,
        r#type: msg.type_.into(),
        markers: msg.markers.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker::interactive_marker_to_bus).collect(),
        poses: msg.poses.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker_pose::interactive_marker_pose_to_bus).collect(),
        erases: crate::ros2_bridge::mappers::convert::string_seq(msg.erases),
    }
}

pub(crate) fn interactive_marker_update_to_ros(bus: crate::visualization_msgs::msg::v1::InteractiveMarkerUpdate) -> ros_env::visualization_msgs::msg::InteractiveMarkerUpdate {
    ros_env::visualization_msgs::msg::InteractiveMarkerUpdate {
        server_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.server_id),
        seq_num: bus.seq_num,
        type_: bus.r#type as _,
        markers: bus.markers.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker::interactive_marker_to_ros).collect(),
        poses: bus.poses.into_iter().map(crate::ros2_bridge::mappers::visualization_msgs::interactive_marker_pose::interactive_marker_pose_to_ros).collect(),
        erases: crate::ros2_bridge::mappers::convert::ros_string_seq(bus.erases),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct VisualizationMsgsInteractiveMarkerUpdateMapper;

impl TypedTopicMapper for VisualizationMsgsInteractiveMarkerUpdateMapper {
    type Ros = ros_env::visualization_msgs::msg::InteractiveMarkerUpdate;
    type Bus = crate::visualization_msgs::msg::v1::InteractiveMarkerUpdate;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(interactive_marker_update_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(interactive_marker_update_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/visualization_msgs/interactive_marker_update.py
"""Generated mapper for `visualization_msgs/msg/InteractiveMarkerUpdate`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.visualization_msgs.interactive_marker import interactive_marker_to_bus, interactive_marker_to_ros
from robot_bus.ros2_bridge.mappers.visualization_msgs.interactive_marker_pose import interactive_marker_pose_to_bus, interactive_marker_pose_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.visualization_msgs.msg.v1 import InteractiveMarkerUpdate as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def interactive_marker_update_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.server_id = str(msg.server_id)
    bus.seq_num = msg.seq_num
    bus.type = msg.type
    bus.markers.extend([interactive_marker_to_bus(x) for x in msg.markers])
    bus.poses.extend([interactive_marker_pose_to_bus(x) for x in msg.poses])
    bus.erases.extend([str(x) for x in msg.erases])
    return bus


def interactive_marker_update_to_ros(bus):
    from visualization_msgs.msg import InteractiveMarkerUpdate as RosMsg

    out = RosMsg()
    out.server_id = str(bus.server_id)
    out.seq_num = bus.seq_num
    out.type = bus.type
    out.markers = [interactive_marker_to_ros(x) for x in bus.markers]
    out.poses = [interactive_marker_pose_to_ros(x) for x in bus.poses]
    out.erases = [str(x) for x in bus.erases]
    return out


class VisualizationMsgsInteractiveMarkerUpdateMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from visualization_msgs.msg import InteractiveMarkerUpdate as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return interactive_marker_update_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return interactive_marker_update_to_ros(bus)
